buidl/taproot.py

```diff
@@ -153,3 +177,147 @@ def parse(cls, b):
         m = (b_len - 33) // 32
         hashes = [b[33 + 32 * i : 65 + 32 * i] for i in range(m)]
         return cls(tapleaf_version, parity, internal_pubkey, hashes)
+
+
+class TapScript(ScriptPubKey):
+    def tap_leaf(self):
+        return TapLeaf(self)
+
+
+class P2PKTapScript(TapScript):
+    def __init__(self, point):
+        super().__init__()
+        if type(point) == S256Point:
+            raw_point = point.bip340()
+        elif type(point) == bytes:
+            raw_point = point
+        else:
+            raise TypeError("To initialize P2PKTapScript, a point is needed")
+        self.commands = [raw_point, 0xAC]
+
+
+class MultiSigTapScript(TapScript):
+    def __init__(self, points, k):
+        super().__init__()
+        if len(points) == 0:
+            raise ValueError("To initialize MultiSigTapScript at least one point")
+        bip340s = sorted([p.bip340() for p in points])
+        self.points = [S256Point.parse_bip340(b) for b in bip340s]
+        self.commands = [bip340s[0], 0xAC]
+        for bip340 in bip340s[1:]:
+            self.commands.extend([bip340, 0xBA])
+        self.commands.extend([number_to_op_code(k), 0x87])
+
+
+class MuSigTapScript(TapScript):
+    def __init__(self, points):
+        super().__init__()
+        if len(points) == 0:
+            raise ValueError("Need at least one public key")
+        bip340s = sorted([p.bip340() for p in points])
+        self.points = [S256Point.parse_bip340(b) for b in bip340s]
+        preimage = b""
+        for b in bip340s:
+            preimage += b
+        self.commitment = sha256(preimage)
+        self.hashes = [big_endian_to_int(sha256(self.commitment + b)) for b in bip340s]
+        self.hash_lookup = {b: h for h, b in zip(self.hashes, bip340s)}
+        points = [h * p for h, p in zip(self.hashes, self.points)]
+        self.point = S256Point.combine(points)
+        self.commands = [self.point.bip340(), 0xAC]
+
+    def get_tweak_point(self, tweak):
+        if self.point.parity:
+            return -1 * self.point + tweak
+        else:
+            return self.point + tweak
+
+    def sign(self, private_key, k, r, sig_hash, tweak=0):
+        tweak_point = self.get_tweak_point(tweak)
+        msg = r.bip340() + tweak_point.bip340() + sig_hash
+        challenge = big_endian_to_int(hash_challenge(msg)) % N
+        h_i = self.hash_lookup[private_key.point.bip340()]
+        c_i = h_i * challenge % N
+        if r.parity == tweak_point.parity:
+            k_real = k
+        else:
+            k_real = -k
+        if self.point.parity == private_key.point.parity:
+            secret = private_key.secret
+        else:
+            secret = -private_key.secret
+        s_i = (k_real + c_i * secret) % N
+        return s_i
```

Comment:
No need to save this and then return it, you can just return it on the previous line

Reply:
That was to debug. Making it work with a tweak was such a pain.